engines/wintermute/base/gfx/opengl/meshx.cpp

```diff
@@ -0,0 +1,498 @@
+			lexer.advanceOnOpenBraces();
+
+			parseNormalCoords(lexer, vertexCoords);
+		} else if (lexer.tokenIsIdentifier("MeshMaterialList")) {
+			lexer.advanceToNextToken();
+			lexer.advanceOnOpenBraces();
+
+			parseMaterials(lexer, faceCount, filename);
+		} else if (lexer.tokenIsIdentifier("Material")) {
+			warning("MeshX::loadFromX global material will not be properly handled");
+			lexer.advanceToNextToken();
+			Material *mat = new Material(_gameRef);
+			mat->loadFromX(lexer, filename);
+			_materials.add(mat);
+		} else if (lexer.tokenIsIdentifier("XSkinMeshHeader")) {
+			lexer.advanceToNextToken();
+			lexer.advanceOnOpenBraces();
+
+			// if any of this is zero, we should have an unskinned mesh
+			int maxSkinWeightsPerVertex = readInt(lexer);
+			int maxSkinWeightsPerFace = readInt(lexer);
+			int boneCount = readInt(lexer);
+
+			_skinnedMesh = boneCount > 0;
+
+			lexer.advanceToNextToken(); // skip semicolon
+		} else if (lexer.tokenIsIdentifier("SkinWeights")) {
+			// but now we certainly should have a skinned mesh
+			_skinnedMesh = true;
+			lexer.advanceToNextToken();
+			lexer.advanceOnOpenBraces();
+
+			parseSkinWeights(lexer);
+		} else if (lexer.tokenIsIdentifier()) {
+			while (!lexer.reachedClosedBraces()) {
+				lexer.advanceToNextToken();
+			}
+
+			lexer.advanceToNextToken(); // skip closed braces
+		} else if (lexer.reachedClosedBraces()) {
+			lexer.advanceToNextToken(); // skip closed braces
+			return true;
+		} else {
+			warning("MeshX::loadFromX unknown token %i encountered", lexer.getTypeOfToken());
+			lexer.advanceToNextToken();
+		}
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::generateMesh() {
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::findBones(FrameNode *rootFrame) {
+	// normal meshes don't have bones
+	if (!_skinnedMesh) {
+		return true;
+	}
+
+	_boneMatrices.resize(skinWeightsList.size());
+
+	for (uint i = 0; i < skinWeightsList.size(); ++i) {
+		FrameNode *frame = rootFrame->findFrame(skinWeightsList[i]._boneName.c_str());
+
+		if (frame) {
+			_boneMatrices[i] = frame->getCombinedMatrix();
+		} else {
+			warning("MeshX::findBones could not find bone %s", skinWeightsList[i]._boneName.c_str());
+		}
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::update(FrameNode *parentFrame) {
+	if (_vertexData == nullptr) {
+		return false;
+	}
+
+	bool res = false;
+
+	// update skinned mesh
+	if (_skinnedMesh) {
+		BaseArray<Math::Matrix4> finalBoneMatrices;
+		finalBoneMatrices.resize(_boneMatrices.size());
+
+		for (uint i = 0; i < skinWeightsList.size(); ++i) {
+			finalBoneMatrices[i] = *_boneMatrices[i] * skinWeightsList[i]._offsetMatrix;
+		}
+
+		float *vertexDataFloat = reinterpret_cast<float *>(_vertexData);
+
+		for (uint32 i = 0; i < _vertexCount; ++i) {
+			vertexDataFloat[i * 8 + 5] = 0.0f;

```

Comment:
* 8 + 5

Looks a bit magic to me without any context or constants to explain precisely why we're shifting like this.

Reply:
Will add some constants for the offsets.